File: snrg_hr/public/js/job_offer.js
function set_company_code(frm) {
	const companyMap = {
		"SNRG Electricals India Private Limited": "SEI",
		"SNRG Wires India Private Limited": "SWI",
	};

	const company = frm.doc.company || "";
	frm.set_value("custom_company_code", companyMap[company] || "SNRG");
}

function set_fy_short(frm) {
	const dateValue = frm.doc.offer_date || frappe.datetime.get_today();
	const d = frappe.datetime.str_to_obj(dateValue);
	if (!d) return;

	let fyStart = d.getFullYear();
	if (d.getMonth() + 1 < 4) {
		fyStart -= 1;
	}

	const fyShort = String(fyStart + 1).slice(-2);
	frm.set_value("custom_fy_short", fyShort);
}

function set_custom_naming_series(frm) {
	const companyCode = frm.doc.custom_company_code || "SNRG";
	const fyShort = frm.doc.custom_fy_short || "";
	if (!fyShort) return;

	frm.set_value("custom_naming_series", `SNRG/HR/OL/${companyCode}/${fyShort}/.###`);
}

function refresh_naming_fields(frm) {
	set_company_code(frm);
	set_fy_short(frm);
	set_custom_naming_series(frm);
}

frappe.ui.form.on("Job Offer", {
	onload(frm) {
		if (frm.is_new()) {
			refresh_naming_fields(frm);
		}
	},

	refresh(frm) {
		if (frm.is_new()) {
			refresh_naming_fields(frm);
		}
	},

	company(frm) {
		refresh_naming_fields(frm);
	},

	offer_date(frm) {
		refresh_naming_fields(frm);
	},

	before_save(frm) {
		refresh_naming_fields(frm);
	},
});

File: snrg_hr/api/job_offer.py
import frappe
from frappe.model.naming import make_autoname
from frappe.utils import getdate, today


COMPANY_CODE_MAP = {
	"SNRG Electricals India Private Limited": "SEI",
	"SNRG Wires India Private Limited": "SWI",
}


def validate(doc, method=None):
	apply_naming_metadata(doc)


def autoname(doc, method=None):
	apply_naming_metadata(doc)
	doc.name = make_autoname(doc.custom_naming_series)


def before_insert(doc, method=None):
	apply_naming_metadata(doc)


def sync_applicant_attachments(doc, method=None):
	if not doc.job_applicant or not doc.name:
		return

	applicant = frappe.db.get_value(
		"Job Applicant",
		doc.job_applicant,
		["resume_attachment", "custom_sales_projection_attachment"],
		as_dict=True,
	)
	if not applicant:
		return

	for file_url in filter(None, [applicant.resume_attachment, applicant.custom_sales_projection_attachment]):
		_attach_existing_file_to_offer(doc, file_url)


def apply_naming_metadata(doc):
	doc.custom_company_code = _get_company_code(doc.company)
	doc.custom_fy_short = _get_fy_short(doc.offer_date)
	doc.custom_naming_series = _build_naming_series(doc.custom_company_code, doc.custom_fy_short)


def _get_company_code(company):
	return COMPANY_CODE_MAP.get(company or "", "SNRG")


def _get_fy_short(offer_date):
	date_obj = getdate(offer_date or today())
	fy_start = date_obj.year
	if date_obj.month < 4:
		fy_start -= 1
	return str(fy_start + 1)[-2:]


def _build_naming_series(company_code, fy_short):
	return f"SNRG/HR/OL/{company_code}/{fy_short}/.###"


def _attach_existing_file_to_offer(doc, file_url):
	if frappe.db.exists(
		"File",
		{
			"attached_to_doctype": "Job Offer",
			"attached_to_name": doc.name,
			"file_url": file_url,
		},
	):
		return

	source_file = frappe.db.get_value(
		"File",
		{"file_url": file_url},
		["file_name", "is_private", "folder"],
		as_dict=True,
	)
	if not source_file:
		return

	frappe.get_doc(
		{
			"doctype": "File",
			"file_name": source_file.file_name,
			"file_url": file_url,
			"is_private": source_file.is_private,
			"folder": source_file.folder or "Home/Attachments",
			"attached_to_doctype": "Job Offer",
			"attached_to_name": doc.name,
		}
	).insert(ignore_permissions=True)
